### ISSUE TITLE
Host Level Event Support

### ISSUE THREAD

**Issue by mathewc** [Azure/azure-webjobs-sdk-script#281]
Consider adding the ability for user code to be triggered on various host level events, e.g. Startup/Shutdown. Here is a case from SO where the user would like to perform a flush operation on static data: https://stackoverflow.com/questions/36760241/intercept-azure-function-host-shutdown-flush-application-insights-telemetryclie.

Let's see if more scenarios pop up requiring something like this, or if there are other better ways to solve these cases. Ideally we don't really want users thinking in terms of the host - we want them to just focus on relatively short lived functions.

**Comment by davidebbo** [comment]
Indeed, thinking in term of host feels odd in a server-less world. We could be scaling from 1 to 10 hosts/instances, and then back down.

Ideally, we want to be in a place where if something needs to be flushed, it just happens automatically as needed.

**Comment by mathewc** [comment]
If you're doing static level caching of things for performance and only doing periodic flushes, you want the ability to quickly flush stuff if the process is going down. It will often not be possible to flush per function invocation. Now, if your background periodic flush is often enough, perhaps in practice you won't miss much, but ideally you want the chance to flush.

**Comment by CrazyTuna** [comment]
@fabiocav add a [new answer](http://stackoverflow.com/a/36793812/4167200) on how to intercept graceful shutdown using a CancellationToken argument in a function. It works fine for me.
If I need to flush or dispose something when the job is shutting down it is enough for the moment.

**Comment by solvingJ** [comment]
In addition to startup/shutdown hooks, I feel it's important to reiterate the mention of global state and advocate more first class support for it.  If I should open up a dedicated feature request on this topic (separate from startup/shutdown hooks), please let me know.

**Comment by grbspltt** [comment]
Any update to this, 4 years and counting?  Looking for node support to setup DB connections at startup, no matter which function is called first. Right now the work around to support this is quite poor.

**Comment by quality-leftovers** [comment]
Also interested in this functionality. I know there are entity functions, but those do not fit every use case.

If no one can detect when the lifetime of a workload ends there can be no efficient flushing.

- I like the idea of [HostEventTrigger("Shutdown")]
- Since there already is FunctionsStartup maybe we could have a FunctionShutdown (that can consume services via DI)?
